### DISCUSSION TITLE
FIREBASE PINTEREST: Adding user data from different users to the public pins they created

### DISCUSSION THREAD

**Discussion by katherinevfry** [nss-evening-cohort-14/evening-client-side#180] [Help Me!] [answered]
Katy Fry
Have you searched the other help tickets? 
I have! 

Describe your issue
I have an "explore" page that shows all public pins from all users. I'd like to show all users which specific user posted what pin. I tried passing the user data to the function that builds the public pins on the dom, but all it does is display the *current* user. Each pin has the correct uid associated with it. I just can't wrap my head around how to access the data for the user who created the pin.

List any error messages that appear
N/A

What have you googled? 
https://firebase.google.com/docs/auth/web/manage-users

Links to code:
https://github.com/katherinevfry/ASSIGNMENT-Firebase-Pinterest/blob/show-user/src/javascripts/components/events/navEvents.js#L16

https://github.com/katherinevfry/ASSIGNMENT-Firebase-Pinterest/blob/show-user/src/javascripts/helpers/showPublicPins.js

**Comment by drteresavasquez** [comment]
Are you capturing user info in a user node on Firebase? If not you'll not be able to do the merge on the data you need to make the comparison or relationship.

**Reply by katherinevfry** [reply]
I'm not! 

I'm guessing the process looks something like creating a post request to firebase and capturing the user data when the user clicks the login button?

**Reply by drteresavasquez** [reply]
Yep! That's it. Then once you have the user info, you can grab the user info and pins using the promise.all. 

If you want their info on their cards, you can use a find/filter to get the user info and then use the spread to merge the user data with the pin data.

**Reply by katherinevfry** [reply]
Thanks, Dr. T! I'll give this a try.

**Reply by katherinevfry** [reply]
I have my user node working!  🎉 The only issue I'm having is that each time the user logs in, a new data entry is posted to firebase. So the UID will be the same on two entries, but the firebase key is different. I'm currently calling my grabUserInfo function in my checkLoginStatus component in the auth. Is there a way to only grab that info once? To check the uid against existing uids and if the uid is unique, only make the post request then? I'm sure the answer is obvious but I'm struggling to work it out.

**Reply by drteresavasquez** [reply]
Awesome. This was what we talked about in the small group. You'll have to check on login if the user already exists and if so, grand their info. If not, run the post new user method. Does that make sense?

**Reply by katherinevfry** [reply]
It does, I think! so something like.... 

```
if (user) {
  getSingleUser //do the stuff
    if (user.firebaseKey.length === 0) {
      createUserInfo // do the stuff
    }
}
```
I'm not sure my logic is quite right. I've been pulling my hair out for several hours now 😆

**Reply by drteresavasquez** [reply]
Well remember the use you get back will be an object right? So you can check if the object is empty so you can't use .length directly on the object. What could you use instead? Look at this code example (ignore the session storage): https://github.com/nss-evening-cohort-13/firebase-cows-2.0/blob/master/src/javascripts/helpers/data/farmerData.js#L20-L38

**Reply by katherinevfry** [reply]
That worked! Thank you so much! Tomorrow I'll take the fun journey of promise.all-ing some stuff. You're the best, Dr. T!

**Reply by drteresavasquez** [reply]
Yay!!! @katherinevfry can you put a quick comment with your answer so we can mark this as completed?

**Comment by katherinevfry** [comment, marked as answer]
In order to create one data entry in firebase per unique user, you need to:
1. write a get request for a single user
2. write Create functionality (post AND patch) that generates a new user object
3. At login status, write some logic that gets the user object. If the user object response data has a length of 0, call your create functionality! 
4. (Oh yeah, don't forget to set up your firebase rules!)
5. Admire your nifty new user node! Data is beautiful.
